src/subscription.js

```diff
@@ -148,15 +148,15 @@ class Subscription extends EventEmitter {
    */
   _whenReadyResolver() {
     this.removeListener('nosub', this._boundWhenReadyRejecter);
-    this._whenReadyResolveFn();
```

Comment:
What code is calling `_reset`? In those cases, is it expected that it calls `whenReady()` again?

Reply:
The only case I found is in `PublicationClient`, [here](https://github.com/mixmaxhq/publication-client/blob/a2d11a110ea444f987407bdf36e1aa97a05f68cf/src/index.js#L137), it's called when [reconnected](https://github.com/mixmaxhq/publication-client/blob/a2d11a110ea444f987407bdf36e1aa97a05f68cf/src/index.js#L102-L115) event happens.
Potentially it can happen that `reconnected` was emitted and right after it `whenReady()` is called for a subscription in other, am I right?
We don't have a specific place where we do it explicitly though.

Reply:
Maybe it makes sense to unsubscribe from `nosub` and `ready` events in `_reset` function? But anyway I believe using `?` is a proper way to handle a variable that might be `null`, if we had TS, it would yell on us in this place without optional chaining :)

Reply:
> Potentially it can happen that reconnected was emitted and right after it whenReady() is called for a subscription in other, am I right?

Can you share a timeline of what's likely happening? Something like:
1. `whenReady()` is called and this._whenReadyResolveFn is set
2. something else happens...

Reply:
Preliminary info:
1. After subscription is created after `publicationClient.subscribe(...)` we can handle the state after it's ready, using `whenReady()` method.
2. If it's subscription is not ready yet (which means that `PublicationClient` is not ready), we subscribe to the local `ready` event inside the subscription instance [here](https://github.com/mixmaxhq/publication-client/blob/a2d11a110ea444f987407bdf36e1aa97a05f68cf/src/subscription.js#L140). At this point `_whenReadyResolveFn` is set a non-null value.
3. We [wait](https://github.com/mixmaxhq/publication-client/blob/a2d11a110ea444f987407bdf36e1aa97a05f68cf/src/subscription.js#L103) until `PublicationClient` is ready. After it emits `ready` event, `Subscription` [handles it and emits](https://github.com/mixmaxhq/publication-client/blob/a2d11a110ea444f987407bdf36e1aa97a05f68cf/src/subscription.js#L166-L174) an internal `ready` event. Which leads to `_whenReadyResolveFn` being called [here](https://github.com/mixmaxhq/publication-client/blob/a2d11a110ea444f987407bdf36e1aa97a05f68cf/src/subscription.js#L149-L152).

What can go wrong:
1. Let's suppose we use `whenReady()`, but subscription is not ready yet, so we subscribe to `ready` event and define `_whenReadyResolveFn`.
2. `PublicationClient` fails and after some time reconnects. We handle `reconnected` event, [reset subscriptions and connect again](https://github.com/mixmaxhq/publication-client/blob/a2d11a110ea444f987407bdf36e1aa97a05f68cf/src/index.js#L114). It [calls](https://github.com/mixmaxhq/publication-client/blob/a2d11a110ea444f987407bdf36e1aa97a05f68cf/src/index.js#L137)`_reset` for every connection. Which sets `_whenReadyResolveFn` of each subscription to `null`.
3. So when `PublicationClient` is ready after reconnection, it [emits](https://github.com/mixmaxhq/publication-client/blob/a2d11a110ea444f987407bdf36e1aa97a05f68cf/src/index.js#L197-L198) `ready` event.
4. We handle it in subscription and [emit](https://github.com/mixmaxhq/publication-client/blob/a2d11a110ea444f987407bdf36e1aa97a05f68cf/src/subscription.js#L171) a local `ready` event.
5. Before reconnection we used `whenReady()` and **didn't clear** subscriptions for local `ready` and `unsub` events when `_reset` was called. So local `ready` event is handled [here](https://github.com/mixmaxhq/publication-client/blob/a2d11a110ea444f987407bdf36e1aa97a05f68cf/src/subscription.js#L150-L151), but `_whenReadyResolveFn` is `null`, as it was set to it in `_reset`.

Reply:
I will update tests to match the scenario that I wrote above. Does it sound good?

Reply:
Great digging. We're getting closer to a fix.

In the "what can go wrong" case, when a subscription is having trouble connecting initially (and takes the _reset codepath), should `whenReady()` be called when it eventually connects?

Reply:
You mean `whenReady()` that was called initially? Should it resolve or not?
According to [this comment](https://github.com/mixmaxhq/publication-client/blob/a2d11a110ea444f987407bdf36e1aa97a05f68cf/src/index.js#L104-L113), if I get it right, it shouldn't. What do you think?

Reply:
I think it should, right? Publications rely on `whenReady()` being called when they first connect. They shouldn't be concerned with whether the first connection was flaky and had to reconnect.

Perhaps the right solution is to ignore 'reconnect' events for a subscription that hasn't had a 'ready' event? Since it's not really a '**re**connect' if it never connected to being with?

Can you try to repro this behavior locally? Perhaps try something like:
1. Stop the publications service locally
2. Open https://app-local.mixmax.com/dashboard/settings/personal/general (that has a publication). It should fail to load.
3. Start the publications service

Maybe if the timing is right, between steps 2 and 3 it will have reconnected.

Reply:
I was able to reproduce! New findings:
- If publications are not working initially, then client won't be initialized. Even after running publications again, nothing will work.
- If publications are working initially. After client was initialized we can get this error in the following way:
  - Stop publications.
  - Move to a page where JS creates a new subscription and uses `whenReady` for to handle subscription's ready state. It will wait until client is reconnected.
  - Run publications.
  - After client is reconnected `whenReady` will resolve and it will call `_whenReadyResolveFn`. But we cleared it just few moments earlier when client was reconnected. That's how this error occurs.

Reply:
> Perhaps the right solution is to ignore 'reconnect' events for a subscription that hasn't had a 'ready' event? Since it's not really a 'reconnect' if it never connected to being with?

I think it makes sense indeed. But I would slightly alter your suggestion. We can't ignore `reconnect`, as for subscriptions created during connection failure, reconnecting has the same meaning as connecting. So we should handle `reconnected` state. Correct me if I'm missing something.

As I see, the problem here is that even for new subscriptions with pending `ready` events, we [call](https://github.com/mixmaxhq/publication-client/blob/a2d11a110ea444f987407bdf36e1aa97a05f68cf/src/subscription.js#L52-L59) `_reset`. It's ok to call `_reset` for those that are already connected and have no unresolved events. My suggestion is before resetting check whether these subscriptions were already initialized or not. Something like this:

```
if (sub._isReady) {
  sub._reset();
}
```

What do you think?

Reply:
Another approach (maybe a better one) — this all happens because there is a pending `whenReady` promise. So maybe we should base on it instead of altering other parts of the code. E.g. we can create a variable `whenReadyPending` and if it's pending, then we don't reset it's handlers to `null` during `_reset`. In this way we can be sure that each time we have a pending `whenReady`, when it will be resolved we won't have `null` values as handlers.

Reply:
Force-pushed with the last proposed solution to make it clear how I would solve this issue: https://github.com/mixmaxhq/publication-client/pull/37/commits/b0988bdf1f5cad9ce1138ebb1b7c6c8b232fd663

Reply:
Great job digging into this! Publications has long been a source of many 'huh?' moments in the product. Thank you for turning a Sentry issue fix into an opportunity to make this really robust. It's time well invested.

I want to be 100% sure we've covered all test scenarios. Can you please test this locally using an isolated instance of publication-client? There's one [here](https://github.com/mixmaxhq/publication-server/blob/master/examples/current/src/index.js) you can work from.

Here's an example of the level of detail we should have. Ideally these would be automated tests, though it's understandable if that is too difficult to do now.

### Test setup
Given the code:

```
import PublicationClient from 'publication-client';

const client = new PublicationClient('/', {});

client.on('connected', () => console.log('connected successfully'));
client.on('reconnected', () => console.log('reconnected successfully'));
client.on('disconnected', () => console.log('disconnected successfully'));

client
  .subscribe('myCollectionUpdates')
  .whenReady()
  .then(() => {
    console.log('myCollectionUpdates is ready!');
  });

client
  .getCollection('myCollection')
  .on('added', (id, fields) => {
    console.log('myCollection added', id, fields);
  })
  .on('changed', (id, changes) => {
    console.log('myCollection changed', id, changes);
  })
  .on('removed', (id) => {
    console.log('myCollection removed', id);
  });
```

and backend code from [here](https://github.com/mixmaxhq/publication-server/blob/master/examples/current/server.js).

### Test scenario: Normal connection

It should log:
```
connected successfully
myCollectionUpdates is ready!
myCollection added
...
```

### Test scenario: reconnected before ready

It should log:
```
reconnected successfully
connected successfully
myCollectionUpdates is ready!
myCollection added
...
```

### Test: reconnected after ready

### Test: reconnected after idle

### Test: stop before ready

### Test: 'nosub'...

(and more tests)

Reply:
I slightly modified your provided sample code so that it includes `nosub` event. Also, I fixed `reconnected` event, it should be handled by subscription instead of publication client. Here is the code that I was using for local tests:

```js
import PublicationClient from 'publication-client';

const client = new PublicationClient('http://localhost:8410', {});

client.on('connected', () => console.log('connected successfully'));
client.on('disconnected', () => console.log('disconnected successfully'));

const subscription = client.subscribe('myCollectionUpdates');

subscription.whenReady().then(() => {
  console.log('myCollectionUpdates is ready!');
});

subscription
  .on('reconnected', () => console.log('myCollectionUpdates reconnected'))
  .on('nosub', () => console.log('myCollectionUpdates nosub'));

client
  .getCollection('myCollection')
  .on('added', (id, fields) => {
    console.log('myCollection added', id, fields);
  })
  .on('changed', (id, changes) => {
    console.log('myCollection changed', id, changes);
  })
  .on('removed', (id) => {
    console.log('myCollection removed', id);
  });

```

Reply:
Tested for cases that you suggested and one more that I could think of. Should be good to go. What do you think, @bradvogel?

Here are the logs for different scenarios.

#### Test 1. Normal scenario

```
connected successfully
myCollectionUpdates is ready!
myCollection added ...
myCollection changed ...
myCollection removed
```

### Test 2. Server fails and reconnects after normal scenario

This is the same as reconnection after idle.

```
connected successfully
myCollectionUpdates is ready!
myCollection added ...
myCollection changed ...
myCollection removed

/* server fails */
disconnected successfully

/* server starts again */
myCollectionUpdates reconnected
connected successfully
```

#### Test 3. Server fails before subscribing to a collection

Server fails and then runs again. To imitate this I delayed subscribing to a collection on client and during this time shut down the server. And then after subscribing was called, started the server again.

This case is also the same as `reconnected` event being called before `ready`. Because of subscription is being created during "failing" server, `whenReady` is called when it's not ready. When connection is back, `reconnected` is called and then `ready`.

I believe this case also covers the one you mentioned as `stop before ready`. We stop the server before `whenReady` is called and before `ready` events occur.

```
connected successfully

/* server fails */
disconnected successfully

/* server starts again */
myCollectionUpdates reconnected
connected successfully
myCollectionUpdates is ready!
myCollection added ...
myCollection changed ...
myCollection removed
```

#### Test 4. Disconnected in the middle of server events

E.g. let's take a case when server emits `added` and then fails (from the sample `publications-server` example).

```
connected successfully
myCollectionUpdates is ready!
myCollection added ...

/* server fails */
disconnected successfully

/* server starts again */
myCollectionUpdates reconnected
connected successfully
myCollection added ...
myCollection changed ...
myCollection removed
```

#### Test 5. `nosub` test

`nosub` event happens when subscription with specified name is not present. To imitate update `myCollectionUpdates` name on server side to something random or remove this subscription.

```
connected successfully
myCollectionUpdates nosub
```

Reply:
Yep, it does! With previous version I received an error, but with new one it works.

Reply:
Awesome! This might fix lots of "blank screen" bugs in the product (for pages that still rely on publications to bootstrap their data)